File: 154. Find Minimum in Rotated Sorted Array II/solution-logn.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// Time: O(logn), worst case is O(n), when nearly all elements are the same
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    if (nums === null || nums.length === 0) {
        return Number.MAX_SAFE_INTEGER;
    }
    
    var min = nums[0];
    var left = 0;
    var right = nums.length - 1;
    
    while (left <= right) {
        var mid = left + parseInt((right - left) / 2);
        
        if (nums[mid] > nums[left]) {
            min = Math.min(min, nums[left]);
            left = mid + 1;
        } else if (nums[mid] < nums[left]) {
            min = Math.min(min, nums[mid]);
            right = mid - 1;
        } else {                                    // where worst case (duplicate numbers) might happen
            min = Math.min(min, nums[left]);
            left++;
        }
    }
    
    return min;
};

File: Leet-034. Search for a Range/solution-logn.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int [] res = {-1, -1};
        if (nums == null || nums.length == 0) {
            return res;
        }
        
        res[0] = findLeft(nums, target);        
        res[1] = findRight(nums, target);
        return res;
    }
    
    public int findLeft(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] >= target) {
                right = mid;
            } else {
                left = mid;
            }
        }
        
        if (nums[left] == target) {
            return left;
        } else if (nums[right] == target) {
            return right;
        } else {
            return -1;
        }
    }
    
    public int findRight(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] <= target) {
                left = mid;
            } else {
                right = mid;
            }
        }
        
        if (nums[right] == target) {
            return right;
        } else if (nums[left] == target) {
            return left;
        } else {
            return -1;
        }
    }
}